lib/dry/schema/key.rb

```diff
@@ -51,7 +51,7 @@ def read(source)
 
       # @api private
       def write(source, target)
-        read(source) { |value| target[coerced_name] = value }
+        read(source) { |value| target[coerced_name] = deep_symbolize_value(value) }
```

Comment:
I'm sorry but I didn't explain things fully! The KeyMap is explicit, in the sense that it is built based on keys defined by a schema, and it uses the keys it knows about to sanitize an input hash. The whole point of this object is to sanitize a hash explicitly, rather than blindly coercing all the keys.

The problem here is that when a macro is used with `:hash` provided as the type and an external schema as the second arg, ie `value(:hash, SomeSchema)` then `SomeSchema` is not used to extend current's schema types, and key map *is built from types* so we end up with a key map that doesn't include `SomeSchema` keys.

Here's a fix that will probably make all your specs pass:

```diff
diff --git lib/dry/schema/macros/value.rb lib/dry/schema/macros/value.rb
index 905fd8c..665fba8 100644
--- lib/dry/schema/macros/value.rb
+++ lib/dry/schema/macros/value.rb
@@ -28,7 +28,9 @@ module Dry
 
             import_steps(schema)
 
-            type(updated_type) unless custom_type? && !current_type.respond_to?(:of)
+            if !custom_type? || array_type?(current_type) || hash_type?(current_type)
+              type(updated_type)
+            end
           end
 
           trace_opts = opts.reject { |key, _| key == :type_spec || key == :type_rule }
```

Reply:
Correction: the above diff will fix only #296 

In case of #315 it's a bit more tricky as composed schemas are represented as dry logic operations, so we need a way to extract schema types from those in `Macros::Each#value`.

Reply:
Now I got it, thanks! 

I tried to implement your solution and it actually required one more change here:

```ruby
def hash_type?(type)
  primitive_inferrer[type].eql?([::Hash])
end
```

to:

```ruby
def hash_type?(type)
   primitive_inferrer[type].include?(::Hash)
end
```
The reason here is that `primitive_inferrer[current_type]` actually returns `[NilClass, Hash]`, instead of `[Hash]`
The above solution solves all problems with #296 and didn't break anything new, but I guess that the `updated_type` should be assigned only the proper side of `"NilClass | Hash"`, do I get it right?

Almost identical case happens with a schema nested under an array, like this:

```ruby
    inner_schema = Dry::Schema.define do
      required(:name).filled(:string)
    end

    schema = Dry::Schema.define do
      required(:users).maybe(:array).each(inner_schema)
    end

   schema.(users: [{name: "John"}]).errors.to_h
   >> {:users=>{0=>{:name=>["is missing"]}}}
```
and this fails for the similar reason, because `primitive_inferrer[current_type]` returns `[NilClass, Array]`, but this case is more tricky, since there is the whole `build_array_type` method.

Do you have any tips about these?

Reply:
@tadeusz-niemiec ah great findings! In case of `maybe` I think we should enforce a maybe type at the `Macros::Maybe#call` level. It could check if the type that was set is already a maybe type, and if it's not - then override it via `type(current_type.optional)`. You could add a helper method to `Macros::Value` too called `maybe_type?` that would simply do `current_type.meta[:maybe].equal?(true)` and use that.

Makes sense?

Reply:
Makes sense and even works! ;)

So, the  #296 is fixed, but there are still problems with: 

1. An optional array with a nested schema (Honestly I don't have a good idea here and what's more - I can't test it properly, because there is a bug, which causes the optional array validation to fail when a `nil` is passed instead of an Array, like this: 

```ruby
     inner_schema = Dry::Schema.define do
      required(:name).filled(:string)
    end

    schema = Dry::Schema.define do
      required(:users).maybe(:array).each(inner_schema)
    end

   schema.(users: nil)
   >> NoMethodError: undefined method `map' for nil:NilClass
```

2.  A schema with composed two params schemas using OR (basically the #306)

To sum up - this PR is ready to be merged and it solves #296, but I also found a new issue (1st point) and the #306 is still unsolved. I can open a new issue if you'd like to.

Reply:
@tadeusz-niemiec thanks for the hard work and sticking with me here :) Could you report an issue about the 1st point?

Re the second point, this is a somewhat difficult problem to solve. Composed schemas (using `OR`) mean that you may have either one set of keys or another in the input. This means `KeyMap` would have to be extended with the ability to check if a given set of keys exist and coerce that.